File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/encryption/secrets/DeviceFileSecretSpec.kt
package stasis.test.client_android.lib.encryption.secrets

import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.shouldBe
import okio.Buffer
import okio.ByteString.Companion.decodeBase64
import okio.Sink
import okio.Source
import okio.buffer
import stasis.client_android.lib.encryption.secrets.DeviceFileSecret
import java.nio.file.Paths

class DeviceFileSecretSpec : WordSpec({
    "A DeviceFileSecret" should {
        val encryptionIv = "kUuYeWjrwqnA93zYCXn2ZC3Pr5Y4srYEcgrR3jP5KtM="
        val encryptionKey = "QBqEu8Kh6iFGpbgYUWADXRfkVa6wUy5w"

        val plaintextData = "some-plaintext-data"
        val encryptedData = "coKEpIHZVHZtPcYuojvMPcOD8S6a2sH4Xg0gPX8UTKz75Ts="

        val fileSecret = DeviceFileSecret(
            file = Paths.get("/tmp/some/file"),
            iv = encryptionIv.decodeBase64()!!,
            key = encryptionKey.decodeBase64()!!
        )

        "support encryption (source)" {
            val actualEncryptedData = fileSecret
                .encryption(Buffer().writeUtf8(plaintextData) as Source)
                .buffer()
                .readByteString()

            actualEncryptedData shouldBe (encryptedData.decodeBase64()!!)
        }

        "support encryption (sink)" {
            val actualEncryptedData = Buffer()

            fileSecret
                .encryption(actualEncryptedData as Sink)
                .buffer()
                .use {
                    it.writeAll(Buffer().writeUtf8(plaintextData))
                }

            actualEncryptedData.readByteString() shouldBe (encryptedData.decodeBase64()!!)
        }

        "support decryption" {
            val actualPlaintextData = fileSecret
                .decryption(Buffer().write(encryptedData.decodeBase64()!!))
                .buffer()
                .readUtf8()

            actualPlaintextData shouldBe (plaintextData)
        }

        "not render its content via toString" {
            fileSecret.toString() shouldBe ("Secret(${fileSecret.javaClass.name})")
        }
    }
})

File: client/src/test/scala/stasis/test/specs/unit/client/encryption/secrets/DeviceFileSecretSpec.scala
package stasis.test.specs.unit.client.encryption.secrets

import java.nio.file.Paths
import org.apache.pekko.actor.ActorSystem
import org.apache.pekko.stream.scaladsl.Source
import org.apache.pekko.util.ByteString
import stasis.client.encryption.secrets.DeviceFileSecret
import stasis.test.specs.unit.AsyncUnitSpec
import stasis.test.specs.unit.client.EncodingHelpers

class DeviceFileSecretSpec extends AsyncUnitSpec with SecretsConfig with EncodingHelpers {
  "A DeviceFileSecret" should "support encryption" in {
    Source
      .single(ByteString(plaintextData))
      .via(fileSecret.encryption)
      .runFold(ByteString.empty)(_ concat _)
      .map { actualEncryptedData =>
        actualEncryptedData should be(encryptedData.decodeFromBase64)
      }
  }

  it should "support decryption" in {
    Source
      .single(encryptedData.decodeFromBase64)
      .via(fileSecret.decryption)
      .runFold(ByteString.empty)(_ concat _)
      .map { actualPlaintextData =>
        actualPlaintextData should be(ByteString(plaintextData))
      }
  }

  it should "not render its content via toString" in {
    fileSecret.toString should be(s"Secret(${fileSecret.getClass.getName})")
  }

  private implicit val system: ActorSystem = ActorSystem(name = "DeviceFileSecretSpec")

  private val encryptionIv = "kUuYeWjrwqnA93zYCXn2ZC3Pr5Y4srYEcgrR3jP5KtM="
  private val encryptionKey = "QBqEu8Kh6iFGpbgYUWADXRfkVa6wUy5w"

  private val plaintextData = "some-plaintext-data"
  private val encryptedData = "coKEpIHZVHZtPcYuojvMPcOD8S6a2sH4Xg0gPX8UTKz75Ts="

  private val fileSecret = DeviceFileSecret(
    file = Paths.get("/tmp/some/file"),
    iv = encryptionIv.decodeFromBase64,
    key = encryptionKey.decodeFromBase64
  )
}
